AdminServer/appscale/admin/push_worker_manager.py

```diff
@@ -0,0 +1,208 @@
+  PID_DIR,
+  CONFIG_DIR
+)
+from appscale.common.monit_app_configuration import create_config_file
+from .utils import ensure_path
+
+# The number of tasks the Celery worker can handle at a time.
+CELERY_CONCURRENCY = 1000
+
+# The directory where Celery configuration files are stored.
+CELERY_CONFIG_DIR = os.path.join(CONFIG_DIR, 'celery', 'configuration')
+
+# The safe memory in MB per Celery worker.
+CELERY_SAFE_MEMORY = 1000
+
+# The directory where Celery persists state.
+CELERY_STATE_DIR = os.path.join('/', 'opt', 'appscale', 'celery')
+
+# The working directory for Celery workers.
+CELERY_WORKER_DIR = os.path.join(CONFIG_DIR, 'celery', 'workers')
+
+# The directory that workers use for logging.
+CELERY_WORKER_LOG_DIR = os.path.join(LOG_DIR, 'celery_workers')
+
+# The time limit of a running task in seconds. Extra time over the soft limit
+# allows it to catch up to interrupts.
+HARD_TIME_LIMIT = 610
+
+# The soft time limit of a running task.
+TASK_SOFT_TIME_LIMIT = 600
+
+# The worker script for Celery to use.
+WORKER_MODULE = 'appscale.taskqueue.push_worker'
+
+
+class ProjectPushWorkerManager(object):
+  """ Manages the Celery worker for a single project. """
+  def __init__(self, zk_client, monit_operator, project_id):
+    """ Creates a new ProjectPushWorkerManager.
+
+    Args:
+      zk_client: A KazooClient.
+      monit_operator: A MonitOperator.
+      project_id: A string specifying a project ID.
+    """
+    self.project_id = project_id
+    self.monit_operator = monit_operator
+    queues_node = '/appscale/projects/{}/queues'.format(project_id)
+    self.watch = zk_client.DataWatch(queues_node, self._update_worker)
+    self.monit_watch = 'celery-{}'.format(project_id)
+
+  @gen.coroutine
+  def update_worker(self, queue_config):
+    """ Updates a worker's configuration and restarts it.
+
+    Args:
+      queue_config: A JSON string specifying queue configuration.
+    """
+    self._write_worker_configuration(queue_config)
+    status = yield self._wait_for_stable_state()
+
+    # Start the worker if it doesn't exist. Restart it if it does.
+    if status == MonitStates.MISSING:
+      command = self.celery_command()
+      env_vars = {'APP_ID': self.project_id, 'HOST': options.login_ip,
+                  'C_FORCE_ROOT': True}
+      pidfile = os.path.join(PID_DIR, 'celery-{}.pid'.format(self.project_id))
+      create_config_file(self.monit_watch, command, pidfile, env_vars=env_vars,
+                         max_memory=CELERY_SAFE_MEMORY)
+      logging.info('Starting push worker for {}'.format(self.project_id))
+      yield self.monit_operator.reload()
+    else:
+      logging.info('Restarting push worker for {}'.format(self.project_id))
+      # Monit behaves unexpectedly with a restart. This hack makes it behave
+      # more smoothly.
+      yield self.monit_operator.send_command(self.monit_watch, 'stop')
+      yield gen.sleep(.5)
+      yield self.monit_operator.send_command(self.monit_watch, 'start')
```

Comment:
From what I've seen, that condition raises a 503, which would be handled here: https://github.com/AppScale/appscale/compare/master...cdonati:update-queue#diff-9f03a29619a12806ac4a4d4b2ba8adcdR212